File: android/src/main/java/com/noirly/alarmkit/events/AlarmEventEmitter.kt
package com.noirly.alarmkit.events

import com.facebook.react.bridge.ReactApplicationContext
import com.noirly.alarmkit.NativeAlarmKitModule
import com.noirly.alarmkit.repository.AlarmRecord

class AlarmEventEmitter(
  private var module: NativeAlarmKitModule?,
) {
  fun attach(module: NativeAlarmKitModule) {
    this.module = module
  }

  fun detach() {
    module = null
  }

  fun emitAlarmFired(record: AlarmRecord) {
    module?.publishAlarmFired(record)
  }

  fun emitAlarmMissedThenFired(record: AlarmRecord) {
    module?.publishAlarmMissedThenFired(record)
  }

  fun emitAlarmsReconciled(records: List<AlarmRecord>) {
    module?.publishAlarmsReconciled(records)
  }

  fun emitSnoozed(record: AlarmRecord) {
    module?.publishSnoozed(record)
  }

  fun emitDismissed(record: AlarmRecord) {
    module?.publishDismissed(record)
  }

  fun emitPermissionsChanged(context: ReactApplicationContext) {
    module?.publishPermissionsChanged(context)
  }
}

File: ios/Events/AlarmEventEmitter.swift
import Foundation

final class AlarmEventEmitter {
  static let shared = AlarmEventEmitter()

  private weak var module: NativeAlarmKit?

  func attach(_ module: NativeAlarmKit) {
    self.module = module
  }

  func detach() {
    module = nil
  }

  func emitAlarmFired(_ record: AlarmRecord) {
    module?.emitOnAlarmFired(AlarmMapper.toDictionary(record))
  }

  func emitAlarmMissedThenFired(_ record: AlarmRecord) {
    module?.emitOnAlarmMissedThenFired(AlarmMapper.toDictionary(record))
  }

  func emitAlarmsReconciled(_ records: [AlarmRecord]) {
    module?.emitOnAlarmsReconciled(records.map { AlarmMapper.toDictionary($0) })
  }

  func emitSnoozed(_ record: AlarmRecord) {
    module?.emitOnSnoozed(AlarmMapper.toDictionary(record))
  }

  func emitDismissed(_ record: AlarmRecord) {
    module?.emitOnDismissed(AlarmMapper.toDictionary(record))
  }

  func emitPermissionsChanged(_ status: [String: Any]) {
    module?.emitOnPermissionsChanged(status)
  }
}
